### PR TITLE
AssignedByValParameterPassParameterByReferenceQuickFix 

### PR DESCRIPTION
**Pull request by BZngr**
Small change to prior PR: Modified PassParameterByRefQuickFix to use a regular expression to replace the ByVal the parameter.

### CODE REVIEW

File: RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs [hunk 1]
```diff
@@ -19,22 +19,19 @@ public PassParameterByReferenceQuickFix(ParserRuleContext context, QualifiedSele
...
+            Regex rgx = new Regex(pattern);
 
-            var parts = parameter.Split(new char[]{' '},2);
-            if (1 != parts.GetUpperBound(0))
-            {
-                return;
-            }
-            parts[0] = parts[0].Replace(Tokens.ByVal, Tokens.ByRef);
-            var newContent = parts[0] + " " + parts[1];
+            var parameter = Context.GetText();
```
Comment (retailcoder) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: The `Context` here should be an `ArgContext`:

    arg : (OPTIONAL whiteSpace)? ((BYVAL | BYREF) whiteSpace)? (PARAMARRAY whiteSpace)? unrestrictedIdentifier (whiteSpace? LPAREN whiteSpace? RPAREN)? (whiteSpace? asTypeClause)? (whiteSpace? argDefaultValue)?;

...which contains everything you need to know to rewrite the tokens.

The regex will *probably* work, but the *ideal*, safest solution would use a `TokenStreamRewriter` to swap the `BYVAL` token with a `BYREF` token, or just to insert a `BYREF` token where it belongs.

Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: Thanks for the feedback.  I can find only one use of TokenStreamRewriter in the Rubberduck code base.  It does not appear to be an example that helps me with this specific use case.  I looked a bit online and am finding Java Antlr4 documentation for TokenStreamRewriter.  Is the Java documentation useful?...or is there some where else you could point me?

Comment (comintern) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: Don't worry about the `TokenStreamRewriter`. You can get the position of the token from the `Context` object.  There's a good example in the `ReplaceObsoleteCommentMarkerQuickFix` - take a look at [this section](https://github.com/rubberduck-vba/Rubberduck/blob/next/RetailCoder.VBE/Inspections/QuickFixes/ReplaceObsoleteCommentMarkerQuickFix.cs#L25-L37). You can even ignore the part about getting the line count - the identifiers you're replacing will always be on a single line.  I usually find it helpful to set a breakpoint, debug a test that uses that code path, and then examine the tree in the locals window.

Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: OK - I've updated to use a strategy similar to comintern's example.  Just for my future reference - I used regular expressions in the prior solution because I found that they were used frequently in the current code base.  Am I to understand that the preference is to use approaches other than regular expressions going forward?  Thanks.

File: RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs [hunk 2]
```diff
@@ -19,21 +20,26 @@ public PassParameterByReferenceQuickFix(ParserRuleContext context, QualifiedSele
...
+            var selection = Selection.Selection;
             var module = Selection.QualifiedName.Component.CodeModule;
-            {
-                var lines = module.GetLines(selection.StartLine, selection.LineCount);
-                var result = lines.Replace(parameter, newParameter);
-                module.ReplaceLine(selection.StartLine, result);
-            }
+
+            var lines = module.GetLines(selection.StartLine, selection.LineCount);
+            var result = lines.Replace(byValParameter, byRefParameter);
```
Comment (comintern) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: I'm not sure how safe `Replace` is here.  You might want to add a test case for something like:
```
Private Sub Foo(ByVal bar As Long, ByRef barTwo As Long)
    bar = barTwo
    Debug.Print bar
End Sub
```

File: RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs (hunk 1 above)
Comment (comintern) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: Yeah, we actually need to clean out the code-base and consolidate\fix a lot of the code rewriting stuff.  My guess is that (outside of the indenter) most of the regular expressions are hold-overs from RD v1.x (which didn't use the same parser).  There are probably _still_ a bunch of code rewriting errors that we simply haven't found yet.

File: RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs (hunk 2 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: I did add a test case with multiple parameters, one of which is the assigned ByVal parameter.  It passed.  There would need to be a duplicate ByVal parameter with the same name in the sub/function signature to cause a problem during Replace.  I think VBA identify the problem with an error if such a signature existed?

Comment (comintern) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: The line relevant to using replace on a line of code is `bar = barTwo`. If it's only replacing `bar` with `xbar`, it can't change that line to `xbar = xbarTwo`, or it breaks it.

Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: oh...that's a different quickfix (which uses Regular Expressions to ensure only the correct variable names are modified).  This quickfix simply changes the signature from ByVal to ByRef - no change to the variable name.

Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: The other quickfix (to use a local copy) does not change the procedure's signature.

Comment (comintern) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: Sorry about that, assumed it was the other one.  In this case, you _do_ have to take multiple lines into consideration.  Consider (the evil but valid):

```
Private Sub Foo(ByVal _
                bar _
                As _
                Long, _
                ByRef _
                barTwo _
                As _
                Long)
```

Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: OK - fair enough!  I now have a unit test for this, and the above evil code does, in fact, fail.  I'll work on getting this corner-case handled as well.

Comment (comintern) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: FWIW, the corner-cases are the ones that will kill you - @ThunderFrame has been beating that into our collective heads since before I started contributing. Don't get the impression that we're criticising - there's just a ton of twisted but legal syntax to account for. Just imagine the absolute worst possible coding habits, and then account for that.  You'll get the hang of it.  :+1:

Comment (retailcoder) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: This. We need a code-rewriting API... like, *20 releases ago!*

We've indeed used regex in lots of places before we started *using* the Antlr parse trees - at some point between 1.4.3 and 2.0 the entire grammar was rewritten, which fixed a *ton* of issues.. but regex lingered, scattered throughout the earlier features.

Basically you need to rewrite an entire logical line of code (which could very well be more than just one *physical* line of code), and the only 100% safe way to do that is, for now, to work with the tokens.

Picture the user writing such weaponized test code:

    Sub DoSomething( _
        ByVal foo As Long, _
        ByVal _
            bar _
        ByRef barbecue
                        )
        bar = barbecue * _
                   bar + foo / barbecue
    End Sub

With the parse tree tokens you can surgically go and edit the *logical* line of code:

```csharp
    var context = Context as ParameterDeclaration;
    Debug.Asser(context != null && context.BYREF() == null);

    if (context.BYVAL() != null)
    {
        // parameter is passed by value
        // BYVAL token start & end positions are known,
        // so we can extract it from the string with then.
    }
    
    // need to insert the BYREF token just before the identifier token
    // use the identifier token's position to determine where to insert it.
````

It's a very "both hands dirty" approach, playing directly with token positions and the code module contents - we definitely need a nice layer of abstraction there.. but until we come up with that API, we need to work at that low abstraction level with the parse trees.

Comment (comintern) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: ...and on the plus side, you didn't open a memory leak that sucked up ~10MB of ram per second. :innocent:

Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: No worries...I definitely understand the importance of the corner cases.  This looks like the kind of VBA code you get when someone writes it on their Apple watch. :)

Comment (Hosch250) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: What happens when you have a parameter like `ByVal ByValName As ByValType`?

Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: Yes! - the xxxByVal named-variable name was the 'corner-case' I discovered (admittedly by accident) that led to my first updates to this QuickFix.  The original code would have failed on comintern's evil code example above as well.  So...progress.

Comment (comintern) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: @Hosch250 has apparently turned to the dark side too.

Comment (ThunderFrame) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]:
```vb
Enum ByRefByVal
  [ByVal = 1,) ByRef ' ByVal ByRef] = 1
  [ByRef ByVal] = [ByVal = 1,) ByRef ' ByVal ByRef]
End Enum


Function ByValByRef(Optional ByVal ByRefByVal = [ByVal = 1,) ByRef ' ByVal ByRef]) As ByRefByVal
  ByValByRef = [ByVal = 1,) ByRef ' ByVal ByRef]
End Function
```

Comment (ThunderFrame) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]:
```vb
Enum ByRefByVal
  [ByVal = 1,) ByRef ' ByVal ByRef] = 1
  [ByRef ByVal] = [ByVal = 1,) ByRef ' ByVal ByRef]
End Enum


Function ByValByRef(Optional ByVal ByRefByVal = [ByVal = 1,) ByRef ' ByVal ByRef]) As ByRefByVal
  ByValByRef = [ByVal = 1,) ByRef ' ByVal ByRef]
End Function
```

Comment (comintern) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: @ThunderFrame - They need a :sob: option in the "Add your reaction" links for that one.

Comment (ThunderFrame) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: @comintern If that emoji is *drilling my eyes out with a dual-headed impact driver*, I think the <kbd>Java</kbd> community has already proposed it to GitHub.

File: RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs (hunk 1 above)
Comment (Vogel612) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: seems like a good place to put a tech-debt issue in ...

**Comment by BZngr** [comment]
I would like to work on this code a bit more given the comments and suggestions received.  Thanks.
